### ISSUE TITLE
Differential backups functionality not working anymore

### ISSUE THREAD

**Issue by darklinkpower** [mtkennerly/ludusavi#360]
### Ludusavi version

v0.24.1

### Operating system

Windows

### Installation method

Standalone

### Description

The functionality is not working correctly anymore. The backups no longer correctly set a parent and its children.

As it can be seen in the manifest, creating a new backup for a game without any, creates it as a children of a non-existent parent. Subsequent backups are also created also as children of the same non-existent parent and as it can be verified, the files with the same hashes are being backed up again unnecessarily in the different children, using more resources in the process.

```yaml
---
name: "Ys VIII: Lacrimosa of DANA"
drives:
  drive-G: "G:"
backups:
  - name: "."
    when: "2024-06-25T11:42:53.693817900Z"
    files: {}
    registry:
      hash: ~
    children:
      - name: backup-20240625T114253Z-diff.zip
        when: "2024-06-25T11:42:53.559190900Z"
        os: windows
        files:
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/adjust.bin":
            hash: 44c8b6befcd476fa89cb39f7ddebb38d6cfd9677
            size: 29
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/config.bin":
            hash: 44586c065aa4d5ecf51730844ddf0873ba0007f1
            size: 73
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/_smemory.dat":
            hash: 46940e3ae74e5a994d67f8288545548d27d8844e
            size: 65540
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steam_autocloud.vdf":
            hash: 1921095146cff11025e0a01ae80c665cdd7a13db
            size: 51
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0001/icon0.jpg":
            hash: 5a03394757b841b6b45cac93aeaa301691670f0a
            size: 76076
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0001/info.txt":
            hash: 49006658b38551929bd5c5d39301b4c737e776a8
            size: 1297
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0001/savedata.dat":
            hash: ff960e0aff6ff6089201473d3b1216335ecbcc33
            size: 819200
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0255/icon0.jpg":
            hash: 2be15957be0db1ac0d436a2363db4b3f06513a32
            size: 18549
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0255/info.txt":
            hash: 00f013f23821c90f3f2165514790b7fe5129de04
            size: 1297
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0255/savedata.dat":
            hash: 714a66189e37e47870d4411893ee961f937733d9
            size: 819200
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/sdmem1/info.dat":
            hash: 892b7c17fe619e02ec8ae1714b6a83fce53f7302
            size: 220
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/sdmem1/memory.dat":
            hash: 4e7f5a5f915c7de4c4c0c4680e8c0a789410ad00
            size: 5816
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/settings.ini":
            hash: 5b1c7c1d229634082f2fe1c3c480f83d0b63506b
            size: 473
        registry: ~
      - name: backup-20240625T114944Z-diff.zip
        when: "2024-06-25T11:49:44.529676100Z"
        os: windows
        files:
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/adjust.bin":
            hash: 44c8b6befcd476fa89cb39f7ddebb38d6cfd9677
            size: 29
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/config.bin":
            hash: 44586c065aa4d5ecf51730844ddf0873ba0007f1
            size: 73
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/_smemory.dat":
            hash: 46940e3ae74e5a994d67f8288545548d27d8844e
            size: 65540
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steam_autocloud.vdf":
            hash: 1921095146cff11025e0a01ae80c665cdd7a13db
            size: 51
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0001/icon0.jpg":
            hash: 5a03394757b841b6b45cac93aeaa301691670f0a
            size: 76076
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0001/info.txt":
            hash: 49006658b38551929bd5c5d39301b4c737e776a8
            size: 1297
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0001/savedata.dat":
            hash: ff960e0aff6ff6089201473d3b1216335ecbcc33
            size: 819200
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0255/icon0.jpg":
            hash: 2be15957be0db1ac0d436a2363db4b3f06513a32
            size: 18549
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0255/info.txt":
            hash: 00f013f23821c90f3f2165514790b7fe5129de04
            size: 1297
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/data0255/savedata.dat":
            hash: 714a66189e37e47870d4411893ee961f937733d9
            size: 819200
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/sdmem1/info.dat":
            hash: 892b7c17fe619e02ec8ae1714b6a83fce53f7302
            size: 220
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/savedata/steamsave/sdmem1/memory.dat":
            hash: f342edad1a44d64a964a20a0867a5e2a37b1e316
            size: 5816
          "G:/Games/PC/Steam/steamapps/common/Ys VIII Lacrimosa of Dana/settings.ini":
            hash: 5b1c7c1d229634082f2fe1c3c480f83d0b63506b
            size: 473
        registry: ~
```

The issue happens when backing up using Ludusavi and also in Ludusavi Playnite with these settings:
![image](https://github.com/mtkennerly/ludusavi/assets/1389286/ef0c1d96-82b3-40a0-8e20-83e810dadb13)

For comparison, this is an older backup manifest from another game that displays the differential backups functionality working correctly, as seen by having a parent and children set correctly:

```yaml
---
name: Resident Evil 2 (2019)
drives:
  drive-C: "C:"
  drive-G: "G:"
backups:
  - name: backup-20230118T145830Z.zip
    when: "2023-01-18T14:58:30.290029700Z"
    files:
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data00-1.bin":
        hash: dbc3fdfb735eea4c368d7e37c6148f31082874ef
        size: 11192
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data000.bin":
        hash: 8e5c1b6dfc5e4347a5852840cad6bfa3d64e642d
        size: 1599756
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data001Slot.bin":
        hash: f97550e072594dc1e901231ae5cace3390c2cbcd
        size: 4839788
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data002Slot.bin":
        hash: 3461e6dbb0023616eac4f4c4503bc8603e6f1642
        size: 4441148
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data003Slot.bin":
        hash: cb11d725853011f62425e577310e6e2b34d8ac3a
        size: 4450688
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data004Slot.bin":
        hash: cdc5ff13c56932300a976e2a25ad08d2d96d59d0
        size: 4588720
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data013Slot.bin":
        hash: 292480f0f1d7e547a8c1512b19e3c11fd7726954
        size: 1858836
      "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data021Slot.bin":
        hash: fae7eb29c06a664d0a6ed1542ff92d9d1c894611
        size: 1104
      "G:/Games/PC/Steam/steamapps/common/RESIDENT EVIL 2  BIOHAZARD RE2/re2_config.ini":
        hash: 154faf0b26536cf98cad98de00b5ad97101d60f9
        size: 774
    registry:
      hash: ~
    children:
      - name: backup-20240313T060005Z.zip
        when: "2024-03-13T06:00:05.075774400Z"
        os: windows
        files:
          "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data00-1.bin":
            hash: 1c3f0c103516653e8f1c8e206758eab5e415a536
            size: 11192
          "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data000.bin":
            hash: 0540e10695df9ce797280349ed7c496fac50ecd9
            size: 4614892
          "G:/Games/PC/Steam/steamapps/common/RESIDENT EVIL 2  BIOHAZARD RE2/re2_config.ini":
            hash: 0b6bd1501e663c1b7f570ea8cd28427935456fb8
            size: 774
        registry: ~
      - name: backup-20240331T122153Z.zip
        when: "2024-03-31T12:21:53.274418500Z"
        os: windows
        files:
          "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data00-1.bin":
            hash: 643e44d8f0b985b8a6390c50d87dfe15c1059018
            size: 11192
          "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data000.bin":
            hash: 0540e10695df9ce797280349ed7c496fac50ecd9
            size: 4614892
          "G:/Games/PC/Steam/steamapps/common/RESIDENT EVIL 2  BIOHAZARD RE2/re2_config.ini":
            hash: 0b6bd1501e663c1b7f570ea8cd28427935456fb8
            size: 774
        registry: ~
      - name: backup-20240413T002348Z.zip
        when: "2024-04-13T00:23:48.016796600Z"
        os: windows
        files:
          "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data00-1.bin":
            hash: 43782ecb922e6078461e2ded3c700d0095590d2d
            size: 11192
          "C:/Program Files (x86)/Steam/userdata/65029263/883710/remote/win64_save/data000.bin":
            hash: 6f03d57dc95477fc9aef5288190e0c9bb9700dc6
            size: 4614796
          "G:/Games/PC/Steam/steamapps/common/RESIDENT EVIL 2  BIOHAZARD RE2/re2_config.ini":
            hash: 0b6bd1501e663c1b7f570ea8cd28427935456fb8
            size: 774
        registry: ~
```

**Comment by mtkennerly** [comment]
Thanks for reporting this. I'm not sure how this got past the unit tests, but I'll try to find the cause and fix this ASAP.

**Comment by mtkennerly** [comment]
Embarrassingly, this has been an issue ever since multi-backup was added 😞

* In an old backup from before multi-backup, the mapping.yaml file would be missing the `backups` field, so Ludusavi would retroactively add a full backup to the list with the right metadata. This part worked fine.
* However, for brand new games that never had a backup, Ludusavi would *also* insert a full backup in the list. Therefore, Ludusavi would see that there was a full backup and decide it must create a differential backup.

I didn't have a unit test that specifically covered the migration for old backups, and I didn't catch it in my personal use since I only use full backups. I've added a unit test for this, although I think I still need to add more robust end-to-end tests to catch these kinds of issues sooner.

I've added some handling to promote the first differential backup into a full backup in mapping.yaml only when the first full backup is otherwise empty. I'll try to make a new release within a couple of days with this fix.
